src/sciline/sphinxext/domain_types.py

```diff
@@ -0,0 +1,153 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+"""A Sphinx extension for improved typehints rendering.
+
+`sphinx-autodoc-typehints <https://github.com/tox-dev/sphinx-autodoc-typehints>`_
+implements formatting for typehints but often results in hard-to-read annotations
+for :class:`typing.NewType`.
+As those are ubiquitous when using Sciline, this extension improves formatting.
+
+Usage
+-----
+
+This extension relies on
+`sphinx-autodoc-typehints <https://github.com/tox-dev/sphinx-autodoc-typehints>`_
+for the core implementation of typehints rendering.
+Install it as well as Sciline and add both to ``extensions`` in your conf.py:
+
+.. code-block:: python
+
+    extensions = [
+        ...,
+        'sphinx-autodoc-typehints',
+        'sciline.sphinxext.domain_types'
+    ]
+
+``sciline.sphinxext.domain_types`` registers a custom ``typehints_formatter``
+used by ``sphinx-autodoc-typehints``, so do not define your own!
+
+Options
+~~~~~~~
+
+- ``sciline_domain_types_prefix`` (default ``''``): Strip this prefix from type names.
+  For example, if a type is defined as ``mypackage.types.SomeType``, setting
+  ``sciline_domain_types_prefix = 'mypackage'`` results in this type being rendered
+  as ``types.SomeType``.
+- ``sciline_domain_types_extra_aliases`` (default ``{}``):
+  ``sciline.sphinxext.domain_types`` can render aliases instead of the regular
+  type names. This is useful, e.g., for abbreviating long types or types defined
+  in internal modules. ``sciline.sphinxext.domain_types`` defines some default
+  aliases for Scipp, e.g., ``{'scipp._scipp.core.DataArray': 'scipp.DataArray', ...}``
+  which means that occurrences of ``scipp._scipp.core.DataArray`` are rendered as
+  ``scipp.DataArray``. Aliases defined in ``sciline_domain_types_extra_aliases``
```

Comment:
Sciline is independent of Scipp. Should we keep this out of the project?

Reply:
Then we have to put it into the template to duplicate it across repos. Shouldn't be a problem, but this here seemed more convenient for our target audience.

Reply:
Is there a problem with putting it into the template? If not, I would suggest to move this there.